File: python/dgp.py
"""Fixed finite-population generator for the national AI validation study."""
from __future__ import annotations

import hashlib
from pathlib import Path
from typing import Any

import numpy as np
import pandas as pd
import yaml
from scipy.special import expit, logit


def load_config(path: str | Path) -> dict[str, Any]:
    with open(path, "r", encoding="utf-8") as handle:
        return yaml.safe_load(handle)


def deterministic_seed(master: int, label: str, size: int, replication: int, language: str) -> int:
    """Map identifiers to a stable 63-bit integer using SHA-256.

    The byte string is ``master|label|size|replication|language``.  The first
    eight digest bytes are interpreted as an unsigned little-endian integer
    and reduced modulo 2^63-1.  This avoids Python's session-dependent hash().
    """
    text = f"{master}|{label}|{size}|{replication}|{language}".encode("utf-8")
    return int.from_bytes(hashlib.sha256(text).digest()[:8], "little") % (2**63 - 1)


def generate_population(config: dict[str, Any]) -> tuple[pd.DataFrame, pd.DataFrame]:
    p = config["population"]
    master = int(config["project"]["master_seed"])
    rng = np.random.default_rng(master)

    h0 = int(p["hospitals_standard"])
    h1 = int(p["hospitals_priority"])
    hospital_stratum = np.r_[np.zeros(h0, dtype=int), np.ones(h1, dtype=int)]
    hospital_count = h0 + h1

    sizes = np.rint(
        rng.lognormal(
            mean=float(p["hospital_size_log_mean"]),
            sigma=float(p["hospital_size_log_sd"]),
            size=hospital_count,
        )
    ).astype(int)
    sizes = np.clip(sizes, int(p["hospital_size_min"]), int(p["hospital_size_max"]))

    sd = np.array(
        [
            float(p["random_effect_sd_q"]),
            float(p["random_effect_sd_y"]),
            float(p["random_effect_sd_a"]),
        ]
    )
    corr = np.array(
        [
            [1.0, float(p["random_effect_corr_q_y"]), float(p["random_effect_corr_q_a"])],
            [float(p["random_effect_corr_q_y"]), 1.0, float(p["random_effect_corr_y_a"])],
            [float(p["random_effect_corr_q_a"]), float(p["random_effect_corr_y_a"]), 1.0],
        ]
    )
    covariance = corr * np.outer(sd, sd)
    if np.min(np.linalg.eigvalsh(covariance)) < -1e-10:
        raise ValueError("The hospital random-effect covariance matrix is not positive semidefinite.")
    random_effects = rng.multivariate_normal(np.zeros(3), covariance, size=hospital_count)

    patient_frames: list[pd.DataFrame] = []
    for hospital_id in range(hospital_count):
        n_h = int(sizes[hospital_id])
        stratum = int(hospital_stratum[hospital_id])

        probability_q = expit(
            logit(float(p["q_baseline_probability"]))
            + float(p["q_priority_log_odds"]) * stratum
            + random_effects[hospital_id, 0]
        )
        q = rng.binomial(1, probability_q, size=n_h)

        probability_y = expit(
            logit(float(p["y_baseline_probability"]))
            + float(p["y_priority_log_odds"]) * stratum
            + float(p["y_hard_log_odds"]) * q
            + random_effects[hospital_id, 1]
        )
        y = rng.binomial(1, probability_y, size=n_h)

        probability_alert_event = expit(
            logit(float(p["sensitivity_baseline_probability"]))
            + float(p["sensitivity_hard_log_odds"]) * q
            + float(p["sensitivity_priority_log_odds"]) * stratum
            + random_effects[hospital_id, 2]
        )
        probability_alert_nonevent = expit(
            logit(float(p["false_alert_baseline_probability"]))
            + float(p["false_alert_hard_log_odds"]) * q
            + float(p["false_alert_priority_log_odds"]) * stratum
            + 0.20 * random_effects[hospital_id, 1]
        )
        alert_probability = np.where(y == 1, probability_alert_event, probability_alert_nonevent)
        a = rng.binomial(1, alert_probability)

        patient_frames.append(
            pd.DataFrame(
                {
                    "hospital_id": hospital_id,
                    "hospital_stratum": stratum,
                    "hard_subgroup": q.astype(int),
                    "outcome": y.astype(int),
                    "alert": a.astype(int),
                }
            )
        )

    population = pd.concat(patient_frames, ignore_index=True)
    population["true_positive"] = population["outcome"] * population["alert"]

    aggregate_rows: list[dict[str, int]] = []
    for hospital_id, data in population.groupby("hospital_id", sort=True):
        row: dict[str, int] = {
            "hospital_id": int(hospital_id),
            "hospital_stratum": int(data["hospital_stratum"].iloc[0]),
            "N": int(len(data)),
            "Q": int(data["hard_subgroup"].sum()),
            "Y": int(data["outcome"].sum()),
            "TP": int(data["true_positive"].sum()),
        }
        for q in (0, 1):
            subgroup = data.loc[data["hard_subgroup"] == q]
            row[f"N_q{q}"] = int(len(subgroup))
            row[f"Y_q{q}"] = int(subgroup["outcome"].sum())
            row[f"TP_q{q}"] = int(subgroup["true_positive"].sum())
        aggregate_rows.append(row)
    hospitals = pd.DataFrame(aggregate_rows).sort_values("hospital_id").reset_index(drop=True)
    return population, hospitals


def target_truths(population: pd.DataFrame) -> pd.DataFrame:
    rows = []
    masks = {
        "national": np.ones(len(population), dtype=bool),
        "easy": population["hard_subgroup"].eq(0).to_numpy(),
        "hard": population["hard_subgroup"].eq(1).to_numpy(),
    }
    for estimand, mask in masks.items():
        data = population.loc[mask]
        y_total = int(data["outcome"].sum())
        tp_total = int(data["true_positive"].sum())
        rows.append(
            {
                "estimand": estimand,
                "target_parameter": tp_total / y_total,
                "eligible_patients": int(len(data)),
                "positive_cases": y_total,
                "event_prevalence": y_total / len(data),
            }
        )
    return pd.DataFrame(rows)


def observation_parameters(population: pd.DataFrame, config: dict[str, Any]) -> pd.DataFrame:
    """Exact finite-population P_O parameters from first-order inclusion probabilities."""
    h0 = int(config["population"]["hospitals_standard"])
    h1 = int(config["population"]["hospitals_priority"])
    truths = target_truths(population).set_index("estimand")["target_parameter"].to_dict()
    rows: list[dict[str, float | str | int]] = []
    masks = {
        "national": np.ones(len(population), dtype=bool),
        "easy": population["hard_subgroup"].eq(0).to_numpy(),
        "hard": population["hard_subgroup"].eq(1).to_numpy(),
    }
    for m in config["designs"]["hospital_sample_sizes"]:
        m = int(m)
        allocations = {
            "N1": None,
            "N2": config["designs"]["proportional_allocations"][str(m)],
            "N3": config["designs"]["enriched_allocations"][str(m)],
            "N4": config["designs"]["enriched_allocations"][str(m)],
            "N5": config["designs"]["enriched_allocations"][str(m)],
        }
        for strategy, allocation in allocations.items():
            if strategy == "N1":
                pi = np.ones(len(population), dtype=float)
            else:
                m0, m1 = (int(allocation[0]), int(allocation[1]))
                pi = np.where(
                    population["hospital_stratum"].to_numpy() == 0,
                    m0 / h0,
                    m1 / h1,
                )
            for estimand, mask in masks.items():
                data = population.loc[mask]
                weights = pi[mask]
                theta_o = float(
                    np.sum(weights * data["true_positive"].to_numpy())
                    / np.sum(weights * data["outcome"].to_numpy())
                )
                rows.append(
                    {
                        "hospital_sample_size": m,
                        "strategy": strategy,
                        "estimand": estimand,
                        "target_parameter": float(truths[estimand]),
                        "observation_parameter": theta_o,
                        "reference_target_discrepancy": theta_o - float(truths[estimand]),
                    }
                )
    return pd.DataFrame(rows)


def patient_srs_cells(population: pd.DataFrame) -> np.ndarray:
    cells = np.zeros((2, 2, 3), dtype=np.int64)
    for h in (0, 1):
        for q in (0, 1):
            data = population.loc[
                (population["hospital_stratum"] == h)
                & (population["hard_subgroup"] == q)
            ]
            cells[h, q, 2] = int(((data["outcome"] == 1) & (data["alert"] == 1)).sum())
            cells[h, q, 1] = int(((data["outcome"] == 1) & (data["alert"] == 0)).sum())
            cells[h, q, 0] = int((data["outcome"] == 0).sum())
    return cells

File: R/dgp.R
suppressPackageStartupMessages({
  library(data.table)
  library(yaml)
  library(digest)
})

load_config <- function(project) {
  yaml::read_yaml(file.path(project, "config", "config.yaml"))
}

stable_seed <- function(master, label, size, replication, language = "R") {
  text <- paste(master, label, size, replication, language, sep = "|")
  hex <- digest::digest(text, algo = "sha256", serialize = FALSE)
  hi <- strtoi(substr(hex, 1, 4), base = 16L)
  lo <- strtoi(substr(hex, 5, 8), base = 16L)
  as.integer((hi * 65536 + lo) %% 2147483646 + 1)
}

load_locked_population <- function(project) {
  fread(file.path(project, "results", "raw", "finite_population.csv.gz"))
}

generate_population_R <- function(config) {
  # Independent R implementation of the DGP. The default chapter run uses the
  # locked finite population created once and stored in results/raw so both
  # languages evaluate exactly the same target population.
  p <- config$population
  set.seed(config$project$master_seed)
  h0 <- p$hospitals_standard
  h1 <- p$hospitals_priority
  H <- h0 + h1
  Hstr <- c(rep(0L, h0), rep(1L, h1))
  sizes <- round(rlnorm(H, p$hospital_size_log_mean, p$hospital_size_log_sd))
  sizes <- pmax(p$hospital_size_min, pmin(p$hospital_size_max, sizes))

  sdv <- c(p$random_effect_sd_q, p$random_effect_sd_y, p$random_effect_sd_a)
  corr <- matrix(c(
    1, p$random_effect_corr_q_y, p$random_effect_corr_q_a,
    p$random_effect_corr_q_y, 1, p$random_effect_corr_y_a,
    p$random_effect_corr_q_a, p$random_effect_corr_y_a, 1
  ), 3, 3, byrow = TRUE)
  Sigma <- corr * outer(sdv, sdv)
  re <- MASS::mvrnorm(H, mu = c(0, 0, 0), Sigma = Sigma)

  out <- vector("list", H)
  for (h in seq_len(H)) {
    n <- sizes[h]
    hs <- Hstr[h]
    pq <- plogis(qlogis(p$q_baseline_probability) + p$q_priority_log_odds * hs + re[h, 1])
    q <- rbinom(n, 1, pq)
    py <- plogis(qlogis(p$y_baseline_probability) + p$y_priority_log_odds * hs + p$y_hard_log_odds * q + re[h, 2])
    y <- rbinom(n, 1, py)
    pa1 <- plogis(qlogis(p$sensitivity_baseline_probability) + p$sensitivity_hard_log_odds * q + p$sensitivity_priority_log_odds * hs + re[h, 3])
    pa0 <- plogis(qlogis(p$false_alert_baseline_probability) + p$false_alert_hard_log_odds * q + p$false_alert_priority_log_odds * hs + 0.20 * re[h, 2])
    a <- rbinom(n, 1, ifelse(y == 1, pa1, pa0))
    out[[h]] <- data.table(
      hospital_id = h - 1L,
      hospital_stratum = hs,
      hard_subgroup = q,
      outcome = y,
      alert = a,
      true_positive = y * a
    )
  }
  rbindlist(out)
}

aggregate_hospitals <- function(pop) {
  base <- pop[, .(
    N = .N,
    Q = sum(hard_subgroup),
    Y = sum(outcome),
    TP = sum(true_positive),
    N_q0 = sum(hard_subgroup == 0),
    Y_q0 = sum(outcome * (hard_subgroup == 0)),
    TP_q0 = sum(true_positive * (hard_subgroup == 0)),
    N_q1 = sum(hard_subgroup == 1),
    Y_q1 = sum(outcome * (hard_subgroup == 1)),
    TP_q1 = sum(true_positive * (hard_subgroup == 1))
  ), by = .(hospital_id, hospital_stratum)]
  setorder(base, hospital_id)
  base
}

target_truths_R <- function(pop) {
  rbindlist(list(
    pop[, .(estimand = "national", target_parameter = sum(true_positive) / sum(outcome), eligible_patients = .N, positive_cases = sum(outcome), event_prevalence = mean(outcome))],
    pop[hard_subgroup == 0, .(estimand = "easy", target_parameter = sum(true_positive) / sum(outcome), eligible_patients = .N, positive_cases = sum(outcome), event_prevalence = mean(outcome))],
    pop[hard_subgroup == 1, .(estimand = "hard", target_parameter = sum(true_positive) / sum(outcome), eligible_patients = .N, positive_cases = sum(outcome), event_prevalence = mean(outcome))]
  ))
}

observation_parameters_R <- function(pop, config) {
  truth <- target_truths_R(pop)
  h0 <- config$population$hospitals_standard
  h1 <- config$population$hospitals_priority
  ans <- list(); k <- 0L
  for (m in config$designs$hospital_sample_sizes) {
    allocations <- list(
      N1 = NULL,
      N2 = config$designs$proportional_allocations[[as.character(m)]],
      N3 = config$designs$enriched_allocations[[as.character(m)]],
      N4 = config$designs$enriched_allocations[[as.character(m)]],
      N5 = config$designs$enriched_allocations[[as.character(m)]]
    )
    for (strategy in names(allocations)) {
      if (strategy == "N1") {
        pi <- rep(1, nrow(pop))
      } else {
        a <- allocations[[strategy]]
        pi <- ifelse(pop$hospital_stratum == 0, a[[1]] / h0, a[[2]] / h1)
      }
      for (e in c("national", "easy", "hard")) {
        idx <- if (e == "national") rep(TRUE, nrow(pop)) else pop$hard_subgroup == ifelse(e == "easy", 0, 1)
        theta_o <- sum(pi[idx] * pop$true_positive[idx]) / sum(pi[idx] * pop$outcome[idx])
        theta_t <- truth[estimand == e, target_parameter]
        k <- k + 1L
        ans[[k]] <- data.table(
          hospital_sample_size = m,
          strategy = strategy,
          estimand = e,
          target_parameter = theta_t,
          observation_parameter = theta_o,
          reference_target_discrepancy = theta_o - theta_t
        )
      }
    }
  }
  rbindlist(ans)
}

patient_cells_R <- function(pop) {
  cells <- array(0L, dim = c(2, 2, 3))
  for (h in 0:1) for (q in 0:1) {
    d <- pop[hospital_stratum == h & hard_subgroup == q]
    cells[h + 1, q + 1, 1] <- sum(d$outcome == 0)
    cells[h + 1, q + 1, 2] <- sum(d$outcome == 1 & d$alert == 0)
    cells[h + 1, q + 1, 3] <- sum(d$outcome == 1 & d$alert == 1)
  }
  cells
}
